### PR TITLE
kernel: fix yielding between tasks with same deadline

### PR DESCRIPTION
**Pull request by jharris-intel**
Previously two tasks with the same deadline and priority would
always have `z_is_t1_higher_prio_than_t2` `true` in both directions.

This is logically inconsistent, and results in `k_yield` not actually
yielding between identical threads.

Signed-off-by: James Harris <james.harris@intel.com>

### CODE REVIEW

File: kernel/sched.c [hunk 1]
```diff
@@ -100,13 +100,22 @@ bool z_is_t1_higher_prio_than_t2(struct k_thread *thread_1,
...
 
 	return false;
 }
 
+bool z_is_t1_higher_prio_than_t2(struct k_thread *thread_1,
+				 struct k_thread *thread_2)
+{
+	const bool ret = is_t1_higher_prio_than_t2(thread_1, thread_2);
+	/* If a>b and b>a, that's a problem */
+	__ASSERT_NO_MSG(!ret || !is_t1_higher_prio_than_t2(thread_2, thread_1));
```
Comment (andyross) [inline kernel/sched.c:115]: FWIW: I'd suggest not putting an assert in the deepest part of the scheduler loop and instead write a test case that shows up the bug you found.

Comment (jharris-intel) [reply kernel/sched.c:115]: Hm. Ok.

I'll have to keep that in mind - I'm used to a much more assert-heavy codebase than Zephyr is.

- [ ] Remove assert.
- [ ] Add yield test case.

Comment (andyross) [reply kernel/sched.c:115]: It's a judgement call I guess.  We have asserts enabled for basically everything running under tests/, the idea is that it's a routine sort of thing you can run with all the time during development.  And doubling the cost of the scheduler priority predicate seems like it's over the line as far as impact goes.  In this case it seems like a "who ran before who" test should be a simple and welcome addition to the regression suite with equal value and zero overhead.

Comment (andyross) [reply kernel/sched.c:115]: FWIW: I'd be happy to approve the patch without the assert or the test.  But the test would be nice. :)

Comment (jharris-intel) [reply kernel/sched.c]: Ah, I get the distinction now.

I'm used to `ASSERT` / `SLOW_ASSERT`, where `SLOW_ASSERT` is exactly what it says on the tin.

You're saying if I understand correctly that you don't have `SLOW_ASSERT` and try to ensure test coverage instead.

File: tests/kernel/sched/deadline/src/main.c [hunk 2]
```diff
@@ -8,9 +8,13 @@
 #include <random/rand32.h>
 
 #define NUM_THREADS 8
-#define STACK_SIZE (256 + CONFIG_TEST_EXTRA_STACKSIZE)
+/* this should be large enough for us
+ * to print a failing assert if necessary
+ */
```
Comment (andyross) [inline tests/kernel/sched/deadline/src/main.c:13]: Nitpicking: I'd skip the comment.  It's fair to assume that the code author thought the requested stack size was large enough :)

**Review by andyross — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

File: kernel/sched.c (hunk 1 above)
Comment (jharris-intel) [reply kernel/sched.c]: As an aside, in my experience asserts / slowasserts are often better than tests for this sort of thing, for a combination of several reasons:

1. Adding a test for this case only catches this one specific case. If someone else e.g. adds another tiebreaker to `is_t1_higher_prio_than_t2` and misses an edge case, the testcases won't catch that. The testcase only catches the specific case of "deadline priority tiebreaker not working".
2. There's a lot of "free" test coverage that you get when you have asserts for invariants. Your context-switch stresstest that someone wrote to test something else entirely that happens to break this invariant? Actually fails.
3. The external visibility of an invariant failing somewhere deep in the bowels of the codebase is often slim-to-none. (E.g. if this had been an issue that only occurred across multiple processors... good luck testing it. The current approach only works because it's a single processor in cooperative mode.)

File: tests/kernel/sched/deadline/src/main.c (hunk 2 above)
Comment (jharris-intel) [reply tests/kernel/sched/deadline/src/main.c:13]: Fair enough.

The comment was intended mainly to be a reminder that the stack size shouldn't just be large enough in the _passing_ case (it is), but also needs to be large enough to handle printing in the _failing_ case too (it wasn't).
